internal/xcrd/crd.go

```diff
@@ -144,53 +111,71 @@ func ForCompositeResourceClaim(xrd *v1.CompositeResourceDefinition) (*extv1.Cust
 	crd.Spec.Names.Categories = append(crd.Spec.Names.Categories, CategoryClaim)
 
 	for i, vr := range xrd.Spec.Versions {
-		crd.Spec.Versions[i] = extv1.CustomResourceDefinitionVersion{
-			Name:                     vr.Name,
-			Served:                   vr.Served,
-			Storage:                  vr.Referenceable,
-			Deprecated:               pointer.BoolDeref(vr.Deprecated, false),
-			DeprecationWarning:       vr.DeprecationWarning,
-			AdditionalPrinterColumns: append(vr.AdditionalPrinterColumns, CompositeResourceClaimPrinterColumns()...),
-			Schema: &extv1.CustomResourceValidation{
-				OpenAPIV3Schema: BaseProps(),
-			},
-			Subresources: &extv1.CustomResourceSubresources{
-				Status: &extv1.CustomResourceSubresourceStatus{},
-			},
-		}
-
-		p, required, err := getProps("spec", vr.Schema)
+		crdv, err := genCrdVersion(vr)
 		if err != nil {
-			return nil, errors.Wrapf(err, errFmtGetProps, "spec")
-		}
-		specProps := crd.Spec.Versions[i].Schema.OpenAPIV3Schema.Properties["spec"]
-		specProps.Required = append(specProps.Required, required...)
-		for k, v := range p {
-			specProps.Properties[k] = v
+			return nil, err
 		}
+		crdv.AdditionalPrinterColumns = append(crdv.AdditionalPrinterColumns, CompositeResourceClaimPrinterColumns()...)
 		for k, v := range CompositeResourceClaimSpecProps() {
-			specProps.Properties[k] = v
-		}
-		crd.Spec.Versions[i].Schema.OpenAPIV3Schema.Properties["spec"] = specProps
-
-		statusP, statusRequired, err := getProps("status", vr.Schema)
-		if err != nil {
-			return nil, errors.Wrapf(err, errFmtGetProps, "status")
+			crdv.Schema.OpenAPIV3Schema.Properties["spec"].Properties[k] = v
 		}
-		statusProps := crd.Spec.Versions[i].Schema.OpenAPIV3Schema.Properties["status"]
-		statusProps.Required = statusRequired
-		for k, v := range statusP {
-			statusProps.Properties[k] = v
-		}
-		for k, v := range CompositeResourceStatusProps() {
-			statusProps.Properties[k] = v
-		}
-		crd.Spec.Versions[i].Schema.OpenAPIV3Schema.Properties["status"] = statusProps
+		crd.Spec.Versions[i] = *crdv
 	}
 
 	return crd, nil
 }
 
+func genCrdVersion(vr v1.CompositeResourceDefinitionVersion) (*extv1.CustomResourceDefinitionVersion, error) {
+	crdv := extv1.CustomResourceDefinitionVersion{
+		Name:                     vr.Name,
+		Served:                   vr.Served,
+		Storage:                  vr.Referenceable,
+		Deprecated:               pointer.BoolDeref(vr.Deprecated, false),
+		DeprecationWarning:       vr.DeprecationWarning,
+		AdditionalPrinterColumns: vr.AdditionalPrinterColumns,
+		Schema: &extv1.CustomResourceValidation{
+			OpenAPIV3Schema: BaseProps(),
+		},
+		Subresources: &extv1.CustomResourceSubresources{
+			Status: &extv1.CustomResourceSubresourceStatus{},
+		},
+	}
+	s, err := parseSchema(vr.Schema)
+	if err != nil {
+		return nil, err
+	}
+	crdv.Schema.OpenAPIV3Schema.Description = s.Description
+
+	xSpec, ok := s.Properties["spec"]
+	if !ok {
+		return nil, errors.Errorf(errFmtGetProps, "spec")
+	}
```

Comment:
We should not return an error if there is no `spec` field in the properties. This is a behavior change. If you check the previous `getProps` function, we're not returning any error if the field doesn't exist.

Reply:
Thanks for noticing that! I'll add a unit test for an empty XRD, ensure the test passes with the previous version of the code, then fix my code so it behaves the same way.